datamapper/changeset.py

```diff
@@ -143,7 +190,9 @@ class Changeset(Generic[T]):
     VALIDATOR_CLASS = MarshmallowValidator
 
     def __init__(self, data: T) -> None:
-        self._wrapped_data: ChangesetDataWrapper[T] = ChangesetDataWrapper(data)
+        self._wrapped_data: ChangesetDataWrapper[T] = ChangesetDataWrapper.instance(
```

Comment:
`T` might be `Tuple[dict, dict]`. In which case, I don't think `ChangesetDataWrapper[T]` is the correct type.

Reply:
So I think I fixed some of the type hints, but I'm left with another problem that I'm not quite sure how to solve because I'm really not that familiar with using generics (or types in general to be honest).

```
datamapper/repo.py:135: error: Item "Dict[Any, Any]" of "Union[Dict[Any, Any], Model]" has no attribute
"__table__"
            sql = model.__table__.insert().values(**changeset.changes)
                  ^
datamapper/repo.py:137: error: Incompatible return value type (got "Union[Dict[Any, Any], Model]",
expected "Union[Model, Changeset[Any]]")
            return changeset.change({"id": record_id}).apply_changes()
                   ^
datamapper/repo.py:153: error: Item "Dict[Any, Any]" of "Union[Dict[Any, Any], Model]" has no attribute
"to_query"
            query = record.to_query()
                    ^
datamapper/repo.py:156: error: Incompatible return value type (got "Union[Dict[Any, Any], Model]",
expected "Union[Model, Changeset[Any]]")
            return changeset.apply_changes()
```

I think the issue here is that `apply_changes()` returns `Data`, which could be `dict` or `Model`. We need to specify somehow that Changeset.apply_changes() returns a model in this case, which I think means that we need to update the TypeVar in the generic. I'm just having some trouble wrapping my head around it.

Reply:
It's hard to get an understanding of these sorts of problems until you dig in for yourself. So, [I took a stab at this here](https://github.com/rzane/datamapper/commit/0eb5ef68786b0974de6d90d4935d2ef8c54b8668).

Here's what I learned:
* A union type in python is actually an intersection type in typescript.
* In order to have this work with 100% accuracy, we'd need conditional types. Otherwise, the changeset's constructor has to relax and allow all of these possible inputs: `Model`, `dict`, `Tuple[dict, dict]`, `Tuple[Model, dict]`. Only two of those are actually valid, but it's *pretty close* to being correct.

Even once the generic type is setup in a way where it should understand that `apply_changes` returns whatever data was passed into the constructor, it turns out that mypy doesn't infer anything.

```python
# no error
Changeset({}).apply_changes().association("jawn")

# error
Changeset[dict]({}).apply_changes().association("jawn")
```

Reply:
Ah, I see. I'm fine with `Union[Data, Tuple[Data, dict]]` if that's the best we can do. As long as the types are passing and providing us some value then I'm cool with it. Thanks for the explanation 👍 

I pulled in your commit and CI should be passing now